### PR TITLE
[Navigation] Register all plugins to NavGroups

### PR DESCRIPTION
**Pull request by TackAdam**
### Description

[Demo.mov.zip](https://github.com/user-attachments/files/16290123/Demo.mov.zip)

Screenshot 2024-07-18 at 4 46 03 PM
Screenshot 2024-07-18 at 4 46 13 PM

This change adds a function that registers all the plugins to their corresponding NavGroups and NavCategories:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/tackadam/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/tackadam/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Plug-in | Group | Category
-- | -- | --
Notebooks | Observability | Dashboards and report
Notebooks | Security Analytics | Dashboards and report
Notebooks | Analytics | Dashboards and report
Integrations | Observability | Dashboards and report
Integrations | Security Analytics | Dashboards and report
Integrations | Analytics | Dashboards and report
Metrics | Observability | Investigate
Applications | Observability | Dashboards and report
Traces | Observability | Investigate

</body>

</html>

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7029

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by joshuali925** [comment]
should this be done inside each plugin? rather than in observability

File: common/constants/shared.ts [hunk 1]
```diff
@@ -58,7 +58,7 @@ export const observabilityNotebookTitle = 'Notebooks';
 export const observabilityNotebookPluginOrder = 5094;
 
 export const observabilityPanelsID = 'observability-dashboards';
-export const observabilityPanelsTitle = 'Dashboards';
```
Comment (ps48) [inline common/constants/shared.ts:61]: can't change this here as it also changes in the name existing side-nav

Comment (RyanL1997) [reply common/constants/shared.ts:61]: +1

File: public/plugin.tsx [hunk 2]
```diff
@@ -330,13 +344,23 @@ export class ObservabilityPlugin
...
     });
 
-    core.application.register({
-      id: observabilityPanelsID,
-      title: observabilityPanelsTitle,
-      category: OBSERVABILITY_APP_CATEGORIES.observability,
-      order: observabilityPanelsPluginOrder,
-      mount: appMountWithStartPage('dashboards'),
-    });
+    if (core.chrome.navGroup.getNavGroupEnabled()) {
```
Comment (RyanL1997) [inline public/plugin.tsx:347]: Just for my knowledge, is this like a feature flag?

**Review by RyanL1997 — APPROVED** [review]

File: public/plugin.tsx (hunk 2 above)
Comment (TackAdam) [reply public/plugin.tsx:347]: Yes, the navigation change can be toggled, and the UI plan currently list regular dashboards as going into the same category. Until it is depreciated the idea was to rename ours for clarity while using the new navigation.

File: public/components/trace_analytics/home.tsx [hunk 3]
```diff
@@ -364,6 +366,7 @@ export const Home = (props: HomeProps) => {
             />
           )}
         />
+        <Route path="/" render={() => <Redirect to={defaultRoute} />} />
```
Comment (joshuali925) [inline public/components/trace_analytics/home.tsx:369]: why this change?

**Comment by TackAdam** [comment]
Latest commit reflects most recent mocks as of 7/17 
While the new navigation is enabled logs, applications, and dashboards are disabled.
Traces is broken into its sections of Traces and Services when it is enabled.

Screenshot 2024-07-17 at 2 36 33 PM

File: public/components/trace_analytics/home.tsx (hunk 3 above)
Comment (SuZhou-Joe) [reply public/components/trace_analytics/home.tsx:369]: Because we need to remove in-app navigation when the new navigation is enabled. In such case, different apps registered to new left navigation will have difference landing pages.

File: public/plugin.tsx [hunk 4]
```diff
@@ -286,26 +291,11 @@ export class ObservabilityPlugin
...
-    });
-
-    core.application.register({
-      id: observabilityLogsID,
-      title: observabilityLogsTitle,
-      category: OBSERVABILITY_APP_CATEGORIES.observability,
-      order: observabilityLogsPluginOrder,
-      mount: appMountWithStartPage('logs'),
-    });
-
```
Comment (SuZhou-Joe) [inline public/plugin.tsx:308]: The existing registration should be there once the new navigation is turned off. Can we keep this part?

File: public/plugin.tsx [hunk 5]
```diff
@@ -314,13 +304,54 @@ export class ObservabilityPlugin
...
+        mount: appMountWithStartPage('traces', '/traces'),
+      });
+
+      core.application.register({
+        id: observabilityApplicationsID,
+        title: observabilityApplicationsTitle,
+        category: OBSERVABILITY_APP_CATEGORIES.observability,
+        order: observabilityApplicationsPluginOrder,
+        mount: appMountWithStartPage('applications'),
+      });
```
Comment (SuZhou-Joe) [inline public/plugin.tsx:322]: This part should be moved out of the if.

File: public/plugin.tsx [hunk 6]
```diff
@@ -314,13 +304,54 @@ export class ObservabilityPlugin
...
+        order: observabilityPanelsPluginOrder,
+        mount: appMountWithStartPage('dashboards'),
+      });
+      core.application.register({
+        id: observabilityLogsID,
+        title: observabilityLogsTitle,
+        category: OBSERVABILITY_APP_CATEGORIES.observability,
+        order: observabilityLogsPluginOrder,
+        mount: appMountWithStartPage('logs'),
+      });
```
Comment (SuZhou-Joe) [inline public/plugin.tsx:353]: This part should be moved out of the if.

File: public/plugin_nav.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+} from '../common/constants/shared';
+import { CoreSetup } from '../../../src/core/public';
+import { AppPluginStartDependencies } from './types';
+import { DEFAULT_NAV_GROUPS, DEFAULT_APP_CATEGORIES } from '../../../src/core/public';
+
+export function registerAllPluginNavGroups(core: CoreSetup<AppPluginStartDependencies>) {
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
+    {
+      id: observabilityApplicationsID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:19]: OSD core has changed the category to visualizeAndReport but keep the `dashboardAndReport` to keep backward compatible. Could you please change to use visualizeAndReport?

File: public/plugin_nav.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      id: observabilityMetricsID,
+      category: DEFAULT_APP_CATEGORIES.investigate,
+      showInAllNavGroup: true,
+    },
+  ]);
+
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
+    {
+      id: observabilityNotebookID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:34]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 400,
```

File: public/plugin_nav.tsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    },
+  ]);
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS[`security-analytics`], [
+    {
+      id: observabilityNotebookID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:40]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 400,
```

File: public/plugin_nav.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      id: observabilityApplicationsID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
+    },
+  ]);
+
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
+    {
+      id: observabilityMetricsID,
+      category: DEFAULT_APP_CATEGORIES.investigate,
+      showInAllNavGroup: true,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:27]:
```suggestion
      showInAllNavGroup: true,
      order: 200,
```

File: public/plugin_nav.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS[`security-analytics`], [
+    {
+      id: observabilityNotebookID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
+    },
+  ]);
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.analytics, [
+    {
+      id: observabilityNotebookID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:46]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 400,
```

File: public/plugin_nav.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    },
+  ]);
+
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
+    {
+      id: observabilityIntegrationsID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:53]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 500,
```

File: public/plugin_nav.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    },
+  ]);
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS[`security-analytics`], [
+    {
+      id: observabilityIntegrationsID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:59]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 500,
```

File: public/plugin_nav.tsx [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    },
+  ]);
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.analytics, [
+    {
+      id: observabilityIntegrationsID,
+      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:65]:
```suggestion
      category: DEFAULT_APP_CATEGORIES.dashboardAndReport,
      order: 500,
```

File: public/plugin_nav.tsx [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+      id: 'observability-traces-nav',
+      category: DEFAULT_APP_CATEGORIES.investigate,
+      showInAllNavGroup: true,
+    },
+    {
+      id: 'observability-services-nav',
+      category: DEFAULT_APP_CATEGORIES.investigate,
+      showInAllNavGroup: true,
+    },
+  ]);
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:80]:
```suggestion
  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
    {
      id: 'observability-traces-nav',
      category: DEFAULT_APP_CATEGORIES.investigate,
      showInAllNavGroup: true,
      order: 300,
    },
    {
      id: 'observability-services-nav',
      category: DEFAULT_APP_CATEGORIES.investigate,
      showInAllNavGroup: true,
      order: 100,
    },
  ]);
```

File: public/plugin.tsx [hunk 16]
```diff
@@ -314,13 +304,54 @@ export class ObservabilityPlugin
...
+    } else {
+      core.application.register({
+        id: observabilityTracesID,
+        title: observabilityTracesTitle,
+        category: OBSERVABILITY_APP_CATEGORIES.observability,
+        order: observabilityTracesPluginOrder,
+        mount: appMountWithStartPage('traces'),
+      });
+      // depreciated in new Nav Groups.
```
Comment (SuZhou-Joe) [inline public/plugin.tsx:339]:
```suggestion
      // deprecated in new Nav Groups.
```
typo

File: public/plugin_nav.tsx [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  observabilityIntegrationsID,
+  observabilityMetricsID,
+  observabilityNotebookID,
+} from '../common/constants/shared';
+import { CoreSetup } from '../../../src/core/public';
+import { AppPluginStartDependencies } from './types';
+import { DEFAULT_NAV_GROUPS, DEFAULT_APP_CATEGORIES } from '../../../src/core/public';
+
+export function registerAllPluginNavGroups(core: CoreSetup<AppPluginStartDependencies>) {
+  core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.observability, [
```
Comment (SuZhou-Joe) [inline public/plugin_nav.tsx:16]: Applications is not needed to be added to nav group as it will be deprecated in new left navigation.

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ps48 — APPROVED** [review]

File: public/components/application_analytics/components/application.tsx [hunk 18]
```diff
@@ -217,17 +214,32 @@ export function Application(props: AppDetailProps) {
...
+        },
+      ]);
+    } else {
+      chrome.setBreadcrumbs([
+        ...parentBreadcrumbs,
+        {
+          text: 'Applications',
+          href: '#/',
+        },
+        {
```
Comment (derek-ho) [inline public/components/application_analytics/components/application.tsx:236]: Can we please create a function like generateBreadcrumbs() that will remove the logic from all of these places and move it to somewhere that can be easily changed in the future if need be?

Comment (TackAdam) [reply public/components/application_analytics/components/application.tsx]: Good call out, thanks added in setNavBreadCrumbs()

**Review by sumukhswamy — APPROVED** [review]

**Review by mengweieric — APPROVED** [review]

**Review by ps48 — APPROVED** [review]
